[Review] Add support for uint32_t and int64_t types for BFS (cpp side)

pgera opened the pull request:
Progress towards https://github.com/rapidsai/cugraph/issues/955

GPUtester commented:
Can one of the admins verify this patch?

seunghwak commented:
Need to add a test case for unsigned and long types in https://github.com/rapidsai/cugraph/blob/branch-0.15/cpp/tests/traversal/bfs_test.cu

pgera commented:
Yes, I'll add that. I have been testing the kernels independently with other large graphs that I have.

pgera commented:
Btw, I also cleaned up some code in sm_utils. There were a few unused shfl wrappers, and they all had the old non-syn variants for old cuda archs. I'm assuming that cuda arch older than 300 is anyway not supported by cugraph. Those can be reverted/added back if they are needed.

seunghwak commented:
> Btw, I also cleaned up some code in sm_utils. There were a few unused shfl wrappers, and they all had the old non-sync variants for old cuda archs. I'm assuming that cuda arch older than 300 is anyway not supported by cugraph. Those can be reverted/added back if they are needed.

Thanks, we support Pascal or later, so it's safe to remove the code.

pgera commented:
I realized that the bitmask logic is overengineered. It is correct, but not necessary. When I implemented it, I implemented it with the assumption that max_unsigned can be a valid vid. However, that cannot happen since num_vertices is also represented by unsigned. So the largest valid vid has to be max - 1, which means that max can be used as the invalid marker. So the original implementation just needs to be changed from -1 to ~(0) to support unsigned. I've marked the PR as WIP. I'll push one more change to simplify the logic.

pgera commented:
This is ready now. I'll do SSSP in a separate PR.

BradReesWork commented:
okay to test

BradReesWork commented:
rerun tests

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

BradReesWork commented:
@pgera The issue with this PR is that it does not address the full end-to-end analytic workflow.  From the Python layer, data will still be specified as int32.  Likewise, data from any of the graph classes will also be int32.

pgera commented:
Yes, I do not use the python layer at all. I made the changes to use cugraph's performance numbers as a comparison point for large graphs in some ongoing work. The python side can be a separate PR, which I can do a bit later once I get some free time. Nothing changes for external users with this PR. Perhaps the PR's title can be renamed to emphasize that this is only the cpp portion.

pgera commented:
Additionally, supporting multiple types at the python layer would require going through all the algorithms to ensure that they all work for multiple types. So I think that fixing all the underlying cpp bits for all the algorithms would anyway be necessary before exposing anything globally.

BradReesWork commented:
ok to test

seunghwak commented on cpp/src/traversal/bfs.cu:
```diff
@@ -474,10 +476,14 @@ void bfs(experimental::GraphCSRView<VT, ET, WT> const &graph,
+  static_assert(std::is_same<VT, ET>::value, "VT and ET should be the same time");
+
+  static_assert(std::is_same<WT, float>::value || std::is_same<WT, double>::value,
+                "Unsupported weight data type, please use float or double");
```
These are more dependent on explicit instantiation rather than fundamental requirements of the algorithm and these will be detected in link time anyway.

Better be

```
static_assert(std::is_integral<VT>::value, "...");
static_assert(std::is_integral<ET>::value, "...");
static_assert(std::is_floating_point<WT>::value, "...");  // actually, this is unnecessary for BFS
```

pgera replied:
My concern was that is_integral is too wide (it includes all widths including 8 and 16), and unless the kernel is really tested to work for all widths, it will likely break if you instantiate it with anything other than what is tested. There are implicit assumptions in places for sizes to be at least 32 bits. One example is the dimx params for block/grid etc. So this assert was to signal the current state of affairs. Same for std::is_same<VT, ET>::value.

seunghwak replied:
OK, I see... If you expect this code to break if sizeof(ET) > sizeof(VT), then, better avoid relaxing the check.

pgera replied:
I'll update the check to assert is_integral, size at least 4 bytes, and VT and ET are the same type. I think the kernels need one additional pass to make them really type agnostic and safe. Need to go through all the arithmetic line by line to check for breakage. There may also be orthogonal overflow issues. One way to weed them out would be to use graphs of sizes at the limit of the type. For eg., things like ceil division (a + K -1)/K will start breaking at that point.

seunghwak commented on cpp/src/traversal/bfs.cu:
```diff
@@ -494,17 +500,48 @@ void bfs(experimental::GraphCSRView<VT, ET, WT> const &graph,
   bfs.traverse(start_vertex);
 }
 
+template void bfs<unsigned, unsigned, float>(
```
add
```
// explicit instantiation
```

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -92,7 +92,7 @@ __global__ void fill_unvisited_queue_kernel(int *visited_bmap,
     if (threadIdx.x == (FILL_UNVISITED_QUEUE_DIMX - 1)) {
       IndexType total               = unvisited_thread_offset + n_unvisited_in_int;
-      unvisited_common_block_offset = atomicAdd(unvisited_cnt, total);
+      unvisited_common_block_offset = traversal::atomicAdd(unvisited_cnt, total);
```
Ideally, traversal::atomicAdd is not specific to traversal and better go to RAFT (e.g. https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/cudart_utils.h)

If it is too much for this PR, better add a comment like
```
// FIXME: this should better go to RAFT
```

pgera replied:
Deferring the move to RAFT for now since RAFT is a separate repo. Added the comment.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -299,8 +299,9 @@ __global__ void main_bottomup_kernel(const IndexType *unvisited,
-    IndexType unvisited_vertex = -1;
+
+    visited_bmap_index[0]      = ~IndexType(0);
+    IndexType unvisited_vertex = ~IndexType(0);
```
See

https://github.com/rapidsai/cugraph/pull/838/files#diff-ba59ce02957abb53c35f1a1929fc0374R682

This is not merged yet (and you can just copy and paste this part and to graph.hpp), but I think this is a better way to handle this.

pgera replied:
Done. Added the invalid_idx type to graph.hpp, and invalid_vertex_id and invalid_edge_id derive from it. Btw, technically, there shouldn't be a reason to use -1 for int. std::numeric_limits<T>::max() should work for all types. The only place it will affect things is possibly at the API level where currently -1 is returned as invalid (for eg., for predecessors). In the future, it could be changed to be std::numeric_limits<T>::max() to make it uniform.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -353,7 +354,7 @@ __global__ void main_bottomup_kernel(const IndexType *unvisited,
       // If we haven't found a parent and there's more edge to check
       if (!found && degree > MAIN_BOTTOMUP_MAX_EDGES) {
-        left_unvisited_off = atomicAdd(left_unvisited_cnt, (IndexType)1);
+        left_unvisited_off = traversal::atomicAdd(left_unvisited_cnt, (IndexType)1);
```
static_cast<IndexType>(1) instead of old C type cast.

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -555,7 +556,7 @@ __global__ void bottom_up_large_degree_kernel(IndexType *left_unvisited,
     for (IndexType i_edge = first_i_edge + logical_lane_id; i_edge < end_i_edge;
          i_edge += BOTTOM_UP_LOGICAL_WARP_SIZE) {
-      IndexType valid_parent = -1;
+      IndexType valid_parent = ~IndexType(0);
```
invalid_vertex_id<IndexType>::value is much more readable than ~IndexType(0)

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -564,7 +565,8 @@ __global__ void bottom_up_large_degree_kernel(IndexType *left_unvisited,
 
-      unsigned int warp_valid_p_ballot = cugraph::detail::utils::ballot((valid_parent != -1));
+      unsigned int warp_valid_p_ballot =
+        cugraph::detail::utils::ballot(valid_parent != ~IndexType(0));
```
invalid_vertex_id<IndexType>::value is much more readable than ~IndexType(0)

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -703,16 +705,21 @@ __global__ void topdown_expand_kernel(
+  if (totaldegree > block_offset) {
+    n_items_per_thread_left =
+      (totaldegree - block_offset + TOP_DOWN_EXPAND_DIMX - 1) / TOP_DOWN_EXPAND_DIMX;
+  }
```
```
IndexType n_items_per_thread_left = total_degree > block_offset ? (totaldegree - block_offset + TOP_DOWN_EXPAND_DIMX - 1) / TOP_DOWN_EXPAND_DIMX : 0;
```

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -703,16 +705,21 @@ __global__ void topdown_expand_kernel(
        block_offset += MAX_ITEMS_PER_THREAD_PER_OFFSETS_LOAD * blockDim.x,
-       n_items_per_thread_left -= MAX_ITEMS_PER_THREAD_PER_OFFSETS_LOAD) {
+       n_items_per_thread_left -=
+       min(n_items_per_thread_left, (IndexType)MAX_ITEMS_PER_THREAD_PER_OFFSETS_LOAD)) {
```
Better use std::min instead of custom min.

pgera replied:
Will require --expt-relaxed-constexpr. Is that OK ? I can change all the mins in the file for both host and device code to std::min.

pgera replied:
Changed all mins to std::min and added --expt-relaxed-constexpr.

seunghwak replied:
Oh... Sorry, I didn't see this is a kernel code :-(

pgera replied:
Ok. So leave device mins as is ?

seunghwak replied:
Yeah... Better leave as the original code. If our codebase currently does not use --expt-relaxed-constexpr, better leave as is.

pgera replied:
Reverted.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -703,16 +705,21 @@ __global__ void topdown_expand_kernel(
+       min(n_items_per_thread_left, (IndexType)MAX_ITEMS_PER_THREAD_PER_OFFSETS_LOAD)) {
     // In this loop, we will process batch_set_size batches
     IndexType nitems_per_thread =
       min(n_items_per_thread_left, (IndexType)MAX_ITEMS_PER_THREAD_PER_OFFSETS_LOAD);
```
Better use std::min instead of custom min.

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -834,8 +841,8 @@ __global__ void topdown_expand_kernel(
-            vec_u[iv]                                    = -1;
-            vec_frontier_degrees_exclusive_sum_index[iv] = -1;
+            vec_u[iv]                                    = ~IndexType(0);
+            vec_frontier_degrees_exclusive_sum_index[iv] = ~IndexType(0);
```
invalid_vertex_id<IndexType>::value is much more readable than ~IndexType(0)

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -844,7 +851,7 @@ __global__ void topdown_expand_kernel(
           IndexType u = vec_u[iv];
           // row_ptr for this vertex origin u
-          vec_row_ptr_u[iv] = (u != -1) ? row_ptr[u] : -1;
+          vec_row_ptr_u[iv] = (u != ~IndexType(0)) ? row_ptr[u] : ~IndexType(0);
```
invalid_vertex_id<IndexType>::value is much more readable than ~IndexType(0)

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -856,12 +863,18 @@ __global__ void topdown_expand_kernel(
+
+            if (edge_mask && !edge_mask[edge]) {
+              // Disabling edge
+              row_ptr_u = ~IndexType(0);
```
invalid_vertex_id<IndexType>::value is much more readable than ~IndexType(0)

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -874,7 +887,7 @@ __global__ void topdown_expand_kernel(
           IndexType v = vec_dest_v[iv];
           vec_v_visited_bmap[iv] =
-            (v != -1) ? previous_bmap[v / INT_SIZE] : (~0);  // will look visited
+            (v != ~IndexType(0)) ? previous_bmap[v / INT_SIZE] : (~int(0));  // will look visited
```
invalid_vertex_id<IndexType>::value is much more readable than ~IndexType(0)

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/bfs_kernels.cuh:
```diff
@@ -889,7 +902,7 @@ __global__ void topdown_expand_kernel(
           int is_visited = vec_v_visited_bmap[iv] & m;
 
-          if (is_visited) vec_frontier_candidate[iv] = -1;
+          if (is_visited) vec_frontier_candidate[iv] = ~IndexType(0);
```
Search for ~IndexType(0) and replace with invalid_vertex_id<IndexType>::value

pgera replied:
Done.

seunghwak commented on cpp/src/traversal/traversal_common.cuh:
```diff
@@ -204,6 +218,19 @@ binsearch_maxle(const IndexType* vec, const IndexType val, IndexType low, IndexT
+template <>
+__device__ __forceinline__ long atomicAdd<long>(long* addr, long val)
+{
+  return ::atomicAdd(reinterpret_cast<unsigned long long*>(addr),
```
Better use int64_t instead of long and better add
static_cast(sizeof(unsigned long long) == sizeof(int64_t), "");

pgera replied:
Done.

pgera commented on cpp/src/converters/COOtoCSR.cu:
```diff
@@ -19,13 +19,36 @@
+template std::unique_ptr<experimental::GraphCSR<unsigned, unsigned, float>>
+coo_to_csr<unsigned, unsigned, float>(
+  experimental::GraphCOOView<unsigned, unsigned, float> const &graph,
+  rmm::mr::device_memory_resource *);
```
Done.

pgera commented on cpp/src/converters/COOtoCSR.cu:
```diff
@@ -19,13 +19,36 @@
 
 namespace cugraph {
 
```
Done.

pgera commented on cpp/src/traversal/bfs.cu:
```diff
@@ -461,6 +461,8 @@ void BFS<IndexType>::clean()
   // the vectors have a destructor that takes care of cleaning
 }
 
```
Done.

pgera commented on cpp/tests/traversal/bfs_test.cu:
```diff
@@ -166,7 +171,7 @@ class Tests_BFS : public ::testing::TestWithParam<BFS_Usecase> {
       // predecessors obtained during the C++ BFS traversal
       VT pred = cugraph_pred[i];  // It could be equal to -1 if the node is never reached
-      if (pred == -1) {
+      if (pred == ~VT(0)) {
```
Done.

BradReesWork commented:
rerun tests

BradReesWork commented:
rerun tests

kaatish commented:
@pgera Do you have any more changes in mind for this PR?

pgera commented:
I'm not sure why some of the CI tests are failing. Was there some known CI issue ? I just merged the upstream branch into mine. So if there were any fixes in the last few days, they should be pulled in now. @kaatish, this PR is done. There should be no impact to users at the moment. The next steps in this direction would be to decouple VT and ET and python bindings, which can be done in follow ups. The bottom up issue (#960) still needs to be fixed, but that is independent.
